src/contrib/talloc.rst

```diff
@@ -0,0 +1,394 @@
+``talloc_free(domain)``.
+
+.. mermaid::
+
+    graph TD
+        domain(domain)
+        sysdb(sysdb)
+        name(name)
+        provider(provider)
+        ldb(ldb)
+        file(file)
+        modules(modules)
+        modules_child(...)
+        rules(rules)
+        rules_child(...)
+        names(names)
+        pattern(pattern)
+        fmt(fmt)
+        re(re)
+        data(data)
+        data_child(...)
+        string(string)
+
+        classDef struct fill:green,stroke:green,stroke-width:1px,fill-opacity:0.2
+        linkStyle default interpolate basis
+        class domain,sysdb,ldb,modules,rules,names,re,data struct
+
+        domain --> name
+        domain --> provider
+        domain --> sysdb
+        domain --> names
+        sysdb --> ldb
+        sysdb --> file
+        ldb --> modules --> modules_child
+        ldb --> rules --> rules_child
+        names --> pattern
+        names --> fmt
+        names --> re
+        re --> data --> data_child
+        re --> string
+
+.. code-tabs::
+
+    .. code-tab:: c
+        :label: talloc
+
+        talloc_free(domain);
+
+    .. code-tab:: c
+        :label: malloc
+
+        /* ... */
+
+        void free_sysdb(struct struct sysdb_ctx *sysdb)
+        {
+            free(sysdb->file);
+            free_ldb(sysdb->ldb);
+            free(ldb);
+        }
+
+        void free_domain(struct sss_domain_info *domain)
+        {
+            free(name);
+            free(provider);
+            free_sysdb(domain->sysdb);
+            free_names(domain->names)
+            free(domain);
+        }
+
+        free_domain(domain);
+
+.. _confdb.h: https://github.com/SSSD/sssd/blob/master/src/confdb/confdb.h#L353
+
+.. note::
+
+    Every pointer created by talloc is a memory context on its own.
+
+Changing the hierarchy
+======================
+
+Ideally, what you want to create is a nice oriented graph where nodes are
+structures and leaves are non-structure elements (e.g. strings). The path must
+follow the natural ordering that comes from the structure definition -- that is
+structures are the parent memory contexts and their fields are their children as
+you can see from previous examples.
+
+Sometimes (e.g. when your function produces an output values) you want to change
+the parent context. You can use ``talloc_steal`` or ``talloc_move`` for this
+operation.
+
+.. code-block:: c
+
+    char *value = talloc_strdup(NULL, "Hello world!");
+    struct record *r;
+
+    r = talloc_zero(NULL, struct record);
+    r->name = talloc_strdup(r, "Hello");
+    r->value = talloc_steal(r, value)
+
+.. seealso::
+
+    There are many functions that can create a new talloc memory context. These
+    functions always take an existing talloc context as an argument that is used
+    as the parent context. For example ``talloc_new``, ``talloc_zero``,
+    ``talloc_zero_array``, ``talloc_strdup``, ``talloc_strndup`` and
+    ``talloc_asprintf``.
```

Comment:
`The newly allocated memory context "hangs" off the existing context` could be useful to mention here, it helps me to think about it this way to visualize the graph.